pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,274 @@
+	}
+
+	// get the verification counter (i.e. the number of times the user has initiated phone verification within
+	// the last 24 hours)
+	verificationCounter := signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey]
+	var counter int
+	dailyLimit := s.config.GetVerificationDailyLimit()
+	if verificationCounter == "" {
+		counter = 0
+	} else {
+		counter, err = strconv.Atoi(verificationCounter)
+		if err != nil {
+			// We shouldn't get an error here, but if we do, we should probably set verification counter to the daily
+			// limit so that we at least now have a valid value
+			log.Error(ctx, err, fmt.Sprintf("error converting annotation [%s] value [%s] to integer, on UserSignup: [%s]",
+				v1alpha1.UserSignupVerificationCounterAnnotationKey,
+				signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey], signup.Name))
+			signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey] = strconv.Itoa(dailyLimit)
+			counter = dailyLimit
+		}
+	}
+
+	// check if counter has exceeded the limit of daily limit - if at limit error out
+	if counter >= dailyLimit {
+		log.Error(ctx, err, fmt.Sprintf("%d attempts made. the daily limit of %d has been exceeded", counter, dailyLimit))
+		initError = errors.NewForbiddenError("daily limit exceeded", "cannot generate new verification code")
+	}
+
+	if initError == nil {
+		// generate verification code
+		verificationCode := generateVerificationCode()
+
+		// set the usersignup annotations
+		signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey] = strconv.Itoa(counter + 1)
+		signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] = verificationCode
+		signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey] = now.Add(
+			time.Duration(s.config.GetVerificationCodeExpiresInMin()) * time.Minute).Format(TimestampLayout)
+
+		// Generate the verification message with the new verification code
+		content := fmt.Sprintf(s.config.GetVerificationMessageTemplate(), verificationCode)
+		client := twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), s.HttpClient)
+		msg, err := client.Messages.SendMessage(s.config.GetTwilioFromNumber(), e164PhoneNumber, content, nil)
+		if err != nil {
+			log.Error(ctx, err, fmt.Sprintf("error while sending, code: %d message: %s", msg.ErrorCode, msg.ErrorMessage))
+
+			// If we get an error here then just die, don't bother updating the UserSignup
+			return errors.NewInternalError(err, "error while sending verification code")
+		}
+	}
+
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, err, "error while updating UserSignup resource")
+		initError = errors.NewInternalError(updateErr, "error while updating UserSignup resource")
+	}
+
+	return initError
+}
+
+func generateVerificationCode() string {
+	buf := make([]byte, codeLength)
+	rand.Read(buf)
```

Comment:
There's no conditions that can cause an error here, so it's not handled.

Reply:
yeah, you are probably right, but since the error is handled also in the example in the goland doc https://golang.org/pkg/crypto/rand/ then I thought that it could be worth doing it here as well.
But it's up to you

Reply:
I'd prefer to leave this out because it would be impossible to test, therefore it couldn't be included in the test coverage.

Reply:
I wouldn't really care about the test coverage in this case - IMHO proper error handling has higher priority

Reply:
It makes sense that the Golang doc example shows error handling, because that is exactly what would be required in a typical scenario.  With the way we are using it though the code is 100% deterministic.  If we want to build in error handling for some impossible scenario, then we would also need to propagate it within the service layer, and then provide special handling for it there along with a specific HTTP return code, all of which will never be called and can't be tested for.  If we can't trust the Go language to act in a deterministic fashion within this limited scope, then we really can't trust it anywhere.

Reply:
Yeah, that make sense. Thanks for this explanation and for the discussion :-) :+1: